### PR TITLE
libcurl: add with_ssl option and deprecate others ssl like options + delete fPIC if shared

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **libcurl/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

closes https://github.com/conan-io/conan-center-index/issues/2898, closes https://github.com/conan-io/conan-center-index/issues/2887, closes https://github.com/conan-io/conan-center-index/issues/2234

Current libcurl recipe is hard to maintain and consume.
- for consumers: The behaviour of the different ssl options is counterintuitive. Having to read conanfile and all the branching to understand which ssl library will be pick at the end is not very user friendly.
- for maintenairs: there are a lot of branching based on ssl options scattered in different methods, leading to issues for some combination of ssl's like options.

Therefore:
- a new option is introduced: `with_ssl`, which aggregates `with_openssl`, `with_winssl`, `darwin_ssl` and `with_wolfssl`
options behaviour.
- `with_openssl`, `with_winssl`, `darwin_ssl` and `with_wolfssl` are deprecated.
- deprecated options still work as expected if user doesn't override default value of `with_ssl` option.
  Reasoning:
  - if user overrides `with_ssl`, it means that he has embraced modifications of recipe options.
  - if user doesn't override `with_ssl` option, it might be because:
     - he just wants the default value of `with_ssl` (so we check that deprecated options were not overriden).
     - he still relies on deprecated options.
- default `with_ssl` option value mimics the old default behaviour:
  - darwinssl on Apple like OS
  - openssl on others OS

### CODE REVIEW

File: recipes/libcurl/all/conanfile.py [hunk 1]
```diff
@@ -76,79 +83,97 @@ def imports(self):
...
+            del self.options.with_libpsl
+            del self.options.with_nghttp2
+            del self.options.with_brotli
 
-        if self.options.with_openssl:
-            # enforce shared linking due to openssl dependency
-            if not tools.is_apple_os(self.settings.os) or not self.options.darwin_ssl:
-                self.options["openssl"].shared = self.options.shared
+    def requirements(self):
+        self.requires("zlib/1.2.11")
```
Comment (a4z) [inline recipes/libcurl/all/conanfile.py:158]: please see 
https://stackoverflow.com/questions/23772816/when-do-i-need-zlib-in-openssl/23829863

I am not sure that compression should be the default

**Comment by a4z** [comment]
when I understand right, it says, if you want a shared libcurl, than you also want a shared openssl lib
this can make sense, you might not want to link the whole ssl lib into libcurl ....this would require change in the build script

**Comment by SpaceIm** [comment]
I've tested with Visual Studio, CMake is more than happy to link static openssl into shared libcurl, so it doesn't make sense to force those options if they are not stricly required, since it prevents to have static openssl in a dependency graph as soon as libcurl is shared.

**Comment by SpaceIm** [comment]
Also: shared libcurl fails with MinGW (at link stage of curl exexutable). I think that it was already broken without this PR, but I need to check if it's not a regression.

**Comment by SpaceIm** [comment]
Shouldn't it be more robust to use `tools.unix_path` instead of things like `self.deps_cpp_info["openssl"].rootpath.replace("\\", "/")` in autoconf arguments?

**Comment by jgsogo** [comment]
Is this the same motivation as the `jpeg` option in #2730 ? I would say that it is the same solution via `provides` and virtual/proxy recipe, wdyt?

**Comment by SpaceIm** [comment]
@jgsogo yes and no :smile: most libs listed in this new `with_ssl` option don't share the same API.
Also this PR now fixes others issues and adds features:
- fix `with_brotli=True`
- add optional zlib
- don't force static or shared "ssl" libs if libcurl static or shared
- remove support of wolfssl if CMake build (Visual Studio) and libcurl < 7.70.0 (because it has been implemented in upstream CMakeLists in version 7.70.0).
- allow using nghttp2 in CMake build
- change the name winssl to schannel (according to upstream modifications in configure.ac and CMakeLists)

**Comment by jgsogo** [comment]
> @jgsogo yes and no 😄 most libs listed in this new `with_ssl` option don't share the same API.

I see, the library handles it inside. Even though having different libraries won't violate the ODR and the binary will link successfully, probably in a project you would want to use the same implementation in all the graph, but not sure if Conan should force it (probably not, if we use the _provides_ + _proxy_, there will be impossible to combine different implementations), so I'm totally fine with this option here.

**Review by jgsogo — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Comment by a4z** [comment]
any idea why this happens for the android build?
```
...
checking for HMAC_Init_ex in -lcrypto... no
checking OpenSSL linking with -ldl... no
checking OpenSSL linking with -ldl and -lpthread... no
configure: OPT_SSL: /Users/a4z/elux/conan/zenv1/.conan/data/openssl/1.1.1h/_/_/package/d90c6937a846904107bf4cebd7a8705bc34c61ab
configure: OPENSSL_ENABLED: 
configure: error: --with-ssl was given but OpenSSL could not be detected
libcurl/7.71.1: 
libcurl/7.71.1: ERROR: Package '2cf31def7946f0f0e9bb3470af6914d793dc878b' build failed
libcurl/7.71.1: WARN: Build folder /Users/a4z/elux/conan/zenv1/.conan/data/libcurl/7.71.1/_/_/build/2cf31def7946f0f0e9bb3470af6914d793dc878b
ERROR: libcurl/7.71.1: Error in build() method, line 191
        self._build_with_autotools()
while calling '_build_with_autotools', line 355
        autotools, autotools_vars = self._configure_autotools()
while calling '_configure_autotools', line 451
        self._autotools.configure(vars=self._autotools_vars, args=configure_args)
        ConanException: Error 1 while executing ./configure '--without-libidn2' '--without-librtmp' '--without-libmetalink' '--without-libpsl' '--with-ssl=/Users/a4z/elux/conan/zenv1/.conan/data/openssl/1.1.1h/_/_/package/d90c6937a846904107bf4cebd7a8705bc34c61ab' '--without-libssh2' '--without-nghttp2' '--with-zlib=/Users/a4z/elux/conan/zenv1/.conan/data/zlib/1.2.11/_/_/package/d90c6937a846904107bf4cebd7a8705bc34c61ab/lib' '--without-brotli' '--disable-shared' '--enable-static' '--disable-ldap' '--prefix=/Users/a4z/elux/conan/zenv1/.conan/data/libcurl/7.71.1/_/_/package/2cf31def7946f0f0e9bb3470af6914d793dc878b' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libexecdir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' '--datarootdir=${prefix}/share' --build=x86_64-apple-darwin --host=aarch64-linux-android
```
openssl is obviously there, the path is given (and correct) and it did work before all those changes

**Comment by SpaceIm** [comment]
@a4z Could you open an issue with your profile, conan version etc?

**Comment by a4z** [comment]
done, 
https://github.com/conan-io/conan-center-index/issues/3121

**Pull request merged**